[WPE] Add a soft limit to the amount of GPU memory BitmapTexturePool can hold

magomez opened the pull request:
#### eff294d622ac02b2267f5093ffe7e384d3f554aa
<pre>
[WPE] Add a soft limit to the amount of GPU memory BitmapTexturePool can hold
<a href="https://bugs.webkit.org/show_bug.cgi?id=271765">https://bugs.webkit.org/show_bug.cgi?id=271765</a>

Reviewed by Carlos Garcia Campos.

Add a soft limit to the amount of memory that BitmapTexturePool can hold. When
the limit is reached, be more aggressive trying to release the stored textures.

* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/platform/graphics/texmap/BitmapTexturePool.cpp:
(WebCore::BitmapTexturePool::BitmapTexturePool):
(WebCore::BitmapTexturePool::acquireTexture):
(WebCore::BitmapTexturePool::scheduleReleaseUnusedTextures):
(WebCore::BitmapTexturePool::releaseUnusedTexturesTimerFired):
(WebCore::BitmapTexturePool::enterLimitExceededModeIfNeeded):
(WebCore::BitmapTexturePool::exitLimitExceededModeIfNeeded):
* Source/WebCore/platform/graphics/texmap/BitmapTexturePool.h:

Canonical link: <a href="https://commits.webkit.org/276997@main">https://commits.webkit.org/276997@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5f9871b46d753953abae0516ed1dd802993ed332

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46389 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25530 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48984 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49064 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42430 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48696 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29891 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22986 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/37858 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46967 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22573 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39992 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19098 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/19938 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/41097 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4434 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/42681 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41461 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50893 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/21394 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/17838 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45092 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/22685 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44020 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10263 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23058 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22384 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/BitmapTexturePool.cpp:
```diff
@@ -31,11 +31,21 @@
 
 BitmapTexturePool::BitmapTexturePool()
     : m_releaseUnusedTexturesTimer(RunLoop::current(), this, &BitmapTexturePool::releaseUnusedTexturesTimerFired)
+    , m_tolerance(releaseUnusedSecondsTolerance)
```
I would call this `m_releaseUnusedSecondsTolerance`

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/BitmapTexturePool.cpp:
```diff
@@ -31,11 +31,21 @@
 BitmapTexturePool::BitmapTexturePool()
     : m_releaseUnusedTexturesTimer(RunLoop::current(), this, &BitmapTexturePool::releaseUnusedTexturesTimerFired)
+    , m_tolerance(releaseUnusedSecondsTolerance)
+    , m_interval(releaseUnusedTexturesTimerInterval)
```
And this `m_releaseUnusedTexturesTimerInterval`

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/BitmapTexturePool.cpp:
```diff
@@ -73,16 +87,54 @@ void BitmapTexturePool::releaseUnusedTexturesTimerFired()
+        m_tolerance = releaseUnusedSecondsToleranceOnLimitExceeded;
+        m_interval = releaseUnusedTexturesTimerIntervalOnLimitExceeded;
+        m_releaseUnusedTexturesTimer.stop();
+        m_releaseUnusedTexturesTimer.startOneShot(0_s);
```
You don't need to stop before startOneShot

magomez committed:
[WPE] Add a soft limit to the amount of GPU memory BitmapTexturePool can hold
https://bugs.webkit.org/show_bug.cgi?id=271765

Reviewed by Carlos Garcia Campos.

Add a soft limit to the amount of memory that BitmapTexturePool can hold. When
the limit is reached, be more aggressive trying to release the stored textures.

* Source/WebCore/PlatformWPE.cmake:
* Source/WebCore/platform/graphics/texmap/BitmapTexturePool.cpp:
(WebCore::BitmapTexturePool::BitmapTexturePool):
(WebCore::BitmapTexturePool::acquireTexture):
(WebCore::BitmapTexturePool::scheduleReleaseUnusedTextures):
(WebCore::BitmapTexturePool::releaseUnusedTexturesTimerFired):
(WebCore::BitmapTexturePool::enterLimitExceededModeIfNeeded):
(WebCore::BitmapTexturePool::exitLimitExceededModeIfNeeded):
* Source/WebCore/platform/graphics/texmap/BitmapTexturePool.h:

Canonical link: https://commits.webkit.org/276997@main
